RealmSwift/Projection.swift

```diff
@@ -0,0 +1,357 @@
+    func set(object: ObjectBase) {
+        storage.objectBase = object
+    }
+    /// :nodoc:
+    public var wrappedValue: Value {
+        get {
+            get(objectBase! as! T)
+        }
+        set {
+            precondition(projectedKeyPath is WritableKeyPath<T, Value>)
+            set(objectBase! as! T, newValue)
+        }
+    }
+    /// Declares a property which is lazily initialized to the type's default value.
+    public init(_ projectedKeyPath: KeyPath<T, Value>) {
+        self.projectedKeyPath = projectedKeyPath
+        self.get = {
+            return $0[keyPath: projectedKeyPath]
+        }
+        self.set = {
+            var ref = $0
+            ref[keyPath: projectedKeyPath as! WritableKeyPath<T, Value>] = $1
+        }
+    }
+}
+
+public protocol Projection: ThreadConfined, RealmCollectionValue {
+    associatedtype Root: ObjectBase
+
+    init()
+    func objectClassName() -> String
+}
+
+/// Projection allows to create a light weight const reflection of the original Realm objects with a minimal effort.
+/// And use them as a model in your application.
+///
+/// Example of usage:
+/// ```
+/// public class Person: Object {
+///     @Persisted var firstName = ""
+///     @Persisted var lastName = ""
+///     @Persisted var address: Address? = nil
+///     @Persisted var friends = List<Person>()
+///     @Persisted var reviews = List<String>()
+/// }
+///
+/// public class Address: EmbeddedObject {
+///     @Persisted var city: String = ""
+///     @Persisted var country = ""
+/// }
+///
+/// struct PersonProjection: Projection {
+///     public typealias Root = Person
+///     public init() {
+///     }
+///     @Projected(\Person.firstName) var firstName
+///     @Projected(\Person.address.city) var homeCity
+///     @Projected(\Person.friends.projectTo.firstName) var firstFriendsName: ProjectedList<String>
+/// }
+/// ```
+public extension Projection {
+
+    init(_ object: Root) {
+        self.init()
+        assign(object)
+    }
+
+    fileprivate subscript(label: String) -> _Projected {
+        Mirror(reflecting: self).descendant(label)! as! _Projected
+    }
+
+    func assign(_ object: Root) {
+        let mirror = Mirror(reflecting: self)
+        for child in mirror.children {
+            if child.value is _Projected {
+                let keyPath = \Self.[child.label!]
+                self[keyPath: keyPath].set(object: object)
+            }
+        }
+    }
+    
+    func objectClassName() -> String {
+        return Root.className()
+    }
+}
+
+public enum ProjectionChange {
+    /**
+     If an error occurs, notification blocks are called one time with a `.error`
+     result and an `NSError` containing details about the error. Currently the
+     only errors which can occur are when opening the Realm on a background
+     worker thread to calculate the change set. The callback will never be
+     called again after `.error` is delivered.
+     */
+    case error(_ error: NSError)
```

Comment:
Why is this an `NSError` instead `Error`?

Reply:
It must be a copy paste from `ObjectChange`